### PR TITLE
Rewrite subTypeInfo to use caching mechanism

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Fix unnecessary DB lookups on sites with no sub types due to cache miss

Before
----------------------------------------
Cache being missed

After
----------------------------------------
Cache not being missed

Technical Details
----------------------------------------
So @totten @seamuslee001 I spotted this one getting a lot of setex in our redis logs & it also seems a good one for me to clarify the 'right' methodology. The issue here is that the change in this PR seems to work & fixes it but I'm no longer sure the best way to clear the cache - currently if you add a type it calls

```
     // reset the cache after adding
    self::subTypeInfo(NULL, FALSE, FALSE, TRUE);
```

This PR flattens the key so we would need to do multiple cache deletes to achieve the same - what is the right approach?

Comments
----------------------------------------

### CODE REVIEW

**Comment by seamuslee001** [comment]
Test fails look related @eileenmcnaughton

**Comment by eileenmcnaughton** [comment]
@seamuslee001 yeah this is more a discussion piece at the moment - I had some questions on it ^^

**Comment by seamuslee001** [comment]
@eileenmcnaughton when you say it flatterns the key what do you mean?

**Comment by eileenmcnaughton** [comment]
So the current function stores a static array 

$cache = ['permutation1' => x, 'permutation2' => y];

If we look at things like OptionGroup::values it saves a value for each option group which is like

Cache->set('permutation1', x)
Cache->set('permutation1', y)

This later feels better to me - except when you want to clear them all - which we would do here if editing contact subtypes

**Comment by totten** [comment]
> This PR flattens the key so we would need to do multiple cache deletes to achieve the same - what is the right approach?

IMHO, adding a new cache-group is probably the simplest way to get the desired effect. You'd probably want to:

* Declare the cache in `Container.php`. 
    * Bonus: The default cache doesn't  specify `withArray=>fast`, but you can specify it on a new group. If I'm reading correctly, that's closer to the previous policy (ie using the composition of a php-local cache plus a redis cache).
* Update `CRM_Utils_System::flushCache()` to also flush this cache.

There are alternatives, of course, e.g.

* If you can easily list all the values of `permutation1`, `permutation2`, etc -- then just delete all of them. (`$cache->deleteMultiple(['permutation1', 'permutation2', ...])`).
* Some cache backends support *tagging* and/or *hierarchies*. The PSR-16 interface doesn't expose this functionality. It might be possible to write or find an extension to PSR-16 that makes it possible, but I'd regard that as a distinct project. (Probably 1-3 days' work.)

**Comment by eileenmcnaughton** [comment]
@totten so I feel like we'll quickly get a lot of containers then? Perhaps that is not a bad thing. We can declare them in extensions too it just feels kinda like a heavy lift

**Comment by eileenmcnaughton** [comment]
"Bonus: The default cache doesn't specify withArray=>fast"

@totten I thought we HAD made it so that we got fast cache mgemt by default?

**Comment by totten** [comment]
@eileenmcnaughton Yeah, there is a bit of boilerplate in declaring a cache.

Something [like this](https://github.com/civicrm/civicrm-core/blob/5.16/Civi/Core/Container.php#L153-L171) mitigates that. We can probably come up with a similar mitigation for [flushCache()](https://github.com/civicrm/civicrm-core/blob/5.16/CRM/Utils/System.php#L1437-L1442) so that `flushCache()` basically clears all `cache.*` services.

The most aggressive response to the boilerplate issue would be to move the cache mgmt out of `Civi\Core\Container` and create a dedicated `Civi\Core\CacheManager` service which could be more Clever. Ex: If you try to access an unrecognized cache, then it creates it implicitly/on-demand per some general policy, but does it in a way were we can still discover/clear out any of those lazily-made caches. We'd lose the *presumption* that all caches are injectable services, but otherwise (with a day or so's work) I think it could be done as a drop-in update.

**Comment by totten** [comment]
>@totten I thought we HAD made it so that we got fast cache mgemt by default?

I think that's true for caches that go through the `CRM_Core_BAO_Cache_Psr16` adapter. But to my reading of `CRM_Utils_Cache` and `Civi\Core\Container`, the system-wide short+long caches don't.

**Comment by eileenmcnaughton** [comment]
hmm - I had thought we got to the point where Civi::cache()->get() WAS going through that adapter - we definitely don't want this (or the already merged groups one) to not be array backed.

**Comment by eileenmcnaughton** [comment]
Discussion on chat 

https://chat.civicrm.org/civicrm/pl/9n5gbbesf3bsjn8dzw5q6bcgyw

short version - this risks having a detrimental effect on performance for Redis/Memcache users without figuring out the fastArray stuff

**Comment by eileenmcnaughton** [comment]
Added 'has-test' because after stepping through the tests that failed when I made a mistake I can see there is pretty solid cover here

**Comment by eileenmcnaughton** [comment]
As an aside - I'm pretty sure that we can drop the $ignoreCache parameter too -will put up a PR to do so if I have a reviewer

Screen Shot 2019-08-02 at 4 17 00 PM

**Comment by eileenmcnaughton** [comment]
After reviewing the test fails I decided to rename the cache contactTypes since it needs to be cleared whenever the contact type cache is cleared (outside tests that should be rare)

I also note that when adding or deleting a contactType we need to

1) clear the cached data
2) rebuild the menus

I kinda wonder if
a) we should have an on_clear callback that rebuilds the menus whenever the cache is cleared & 
b) move both to some kinda listener.. - I think @colemanw sees listeners / some sort of post action as the way to go with creates & adds

Also I think we should add a pseudoconstant cache @seamuslee001 to move the cached vars in the Core_Pseudoconstant class into

**Comment by eileenmcnaughton** [comment]
@seamuslee001 I think this is mergeable now

File: CRM/UF/Page/ProfileEditor.php [hunk 1]
```diff
@@ -50,7 +50,7 @@ public static function registerProfileScripts() {
             'phoneType' => CRM_Core_PseudoConstant::get('CRM_Core_DAO_Phone', 'phone_type_id'),
           ],
           'initialProfileList' => $ufGroups,
-          'contactSubTypes' => CRM_Contact_BAO_ContactType::subTypes(),
+          'contactTypes' => CRM_Contact_BAO_ContactType::subTypes(),
```
Comment (seamuslee001) [inline CRM/UF/Page/ProfileEditor.php]: @eileenmcnaughton why the change of key here looks odd

File: CRM/Contact/BAO/ContactType.php [hunk 2]
```diff
@@ -617,8 +599,9 @@ public static function del($contactTypeId) {
 FROM civicrm_navigation
 WHERE name = %1";
       $params = [1 => ["New $name", 'String']];
-      $dao = CRM_Core_DAO::executeQuery($sql, $params);
+      CRM_Core_DAO::executeQuery($sql, $params);
       CRM_Core_BAO_Navigation::resetNavigation();
+      Civi::cache('contactTypes')->clear();
```
Comment (seamuslee001) [inline CRM/Contact/BAO/ContactType.php]: why are you using `clear()` here when we have consistently used `flush()` which is an alias of clear

File: CRM/Contact/BAO/ContactType.php [hunk 3]
```diff
@@ -680,9 +663,7 @@ public static function add(&$params) {
       CRM_Core_BAO_Navigation::add($navigation);
     }
     CRM_Core_BAO_Navigation::resetNavigation();
-
-    // reset the cache after adding
-    self::subTypeInfo(NULL, FALSE, FALSE, TRUE);
+    Civi::cache('contactTypes')->clear();
```
Comment (seamuslee001) [inline CRM/Contact/BAO/ContactType.php]: same as L604

File: CRM/Utils/System.php [hunk 4]
```diff
@@ -1442,6 +1442,7 @@ public static function flushCache() {
       Civi::cache('groups')->flush();
       Civi::cache('navigation')->flush();
       Civi::cache('customData')->flush();
+      Civi::cache('contactTypes')->clear();
```
Comment (seamuslee001) [inline CRM/Utils/System.php]: Same as on ContactType BAO L604

File: CRM/Contact/BAO/ContactType.php (hunk 2 above)
Comment (eileenmcnaughton) [reply CRM/Contact/BAO/ContactType.php]: because flush is actually deprecated!

Comment (seamuslee001) [reply CRM/Contact/BAO/ContactType.php]: ok do we want consistance or non deprecated code paths :-P

Comment (eileenmcnaughton) [reply CRM/Contact/BAO/ContactType.php]: we get consistency when the others are migrated over

File: CRM/UF/Page/ProfileEditor.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/UF/Page/ProfileEditor.php]: @seamuslee001 agree it looks wrong - changed it back

**Comment by seamuslee001** [comment]
ok i'm ok with this now and if Jenkins is then we can merge. I'm pretty confident we have good test coverage covering this

**Comment by eileenmcnaughton** [comment]
@seamuslee001 thanks - working through this makes me feel like we probably also want a cache for 'pseudoconstants' - basically  anything handled by CRM_Core_Pseudoconstant - 

It was a good exercise to work through
